### ISSUE TITLE
task scheduler that support parallelization and crash recovery (continue with only unfinished job and dont rerun existing results)

### ISSUE THREAD

**Issue by smilesun** [marrlab/DomainLab#81]
reference solution: https://github.com/marrlab/DomainLab/pull/79/files

- What happens if the code crashes? Could you still retrieve partially the results instead of letting the whole benchmark to restart again? 

- How do you read a yml file and according to the configuration there execute the benchmark?

- How do you paralllelize the experiment for stochastic variations and hyper-parameter variations? 

snakemake support MPI, slurm, and other task dispatcher

https://snakemake.readthedocs.io/en/stable/executing/cluster.html

**Comment by schoersch** [comment]
I think the easiest way for parallelization would be to run different algorithms alongside each other.

**Comment by rahulbshrestha** [comment]
Based on the slurm scripts, a possible yaml file to represent benchmarking could be:

```
task_1:
	aname: diva
	epos: 1	
	batchsize:
	startseed:
	endseed:
	paths:
		tpath: examples/tasks/demo_task_path_list_small.py
		npath: examples/nets/resnet.py
	algo_conf:
		npath_dom: examples/nets/resnet.py
		gamma_y: 1e5
		gamma_d: 1e5

task_2:
	aname: jigen
	epos: 1
…….
```

This can be combined with the yaml format @schoersch created in https://github.com/marrlab/DomainLab/pull/90

**Comment by smilesun** [comment]
@schoersch @rahulbshrestha @Car-la-F @xinyuejohn 

snakemake support MPI, slurm, and other task dispatcher

https://snakemake.readthedocs.io/en/stable/executing/cluster.html

**Comment by schoersch** [comment]
@rahulbshrestha 
I think it would be easier if we replace
```
	algo_conf:
		npath_dom: examples/nets/resnet.py
		gamma_y: 1e5
		gamma_d: 1e5
```
by
```
	npath_dom: examples/nets/resnet.py
        parameters:
		gamma_y: 1e5
		gamma_d: 1e5
```
such that we have one node containing *only* parameters

**Comment by schoersch** [comment]
I'm going to start with snakemake for task setup and scaling and see how far we get. But there is at least stated:

_"[Workflows] can be seamlessly scaled to server, cluster, grid and cloud environments, without the need to modify the workflow definition."_
[snakemake](https://snakemake.readthedocs.io/en/stable/)

**Comment by schoersch** [comment]
I changed the format of the yaml file a bit. The current yaml file looks like this:

```
# Example benchmark config script.

# Arguments set here are passed to each task.
# Specifications inside the tasks taking precedence.

tpath: examples/tasks/task_vlcs.py  # dataset

# list of all domains that are used as test domain
# in a leave-one-out setup, i.e. for each run,
# one domain from this list is chosen as test domain
# while training is performed on all other domains
# of the specified dataset.
# TODO domains is a misleading name here.
domains:
  - caltech
  # - sun
  # - labelme

output_dir: zoutput/benchmarks/demo_benchmark

# number of hyperparameter samples per task.
# Thus, the total runs of each task are given
# by len(domains) * num_param_samples * num_seeds (see below)
num_param_samples: 2

epos: 2
batchsize: 2
# the seed is increased by +1 until it reaches endseed.
# endseed is included, so in total startseed - endseed + 1
# different seeds are used to estimate the stochastic
# variance.
startseed: 1
endseed: 2  # currently included

# Each node containing the aname property is considered as a task.
Task1:  # name
  # Parameters that are fixed for all runs of this task.
  aname: diva
  nname: alexnet
  nname_dom: alexnet

  # specification of parameters that shall vary
  # between runs to analyze the sensitivity
  # of this task w.r.t. these parameters.
  hyperparameters:
    # Each parameter must contain:
    # - distribution (uniform | loguniform | normal | lognormal)
    # - min and max if distribution is uniform or loguniform
    # - mean and std if distribution is normal or lognormal

    # step is optional and defines discrete parameters
    # with the given step size.
    # If min/mean and step are integer valued,
    # the hyperparameter is ensured to be integer valued too.
    # Otherwise, it is a float and rounding errors can occur.
    gamma_y:
      min: 10e4
      max: 10e6
      step: 1
      distribution: uniform

    gamma_d:
      min: 1e4
      max: 1e6
      step: 1
      distribution: uniform

Task2:
...
```

The snakemake file is design to receive this benchmark yaml file as configfile. As of now, the snakemake than samples the hyperparameters, runs the experiments and aggregates the results in a single csv file.

**Comment by schoersch** [comment]
I spent some time with snakemake and got an early wip state.

@smilesun this adds snakemake as dependency. I'm going to make a pr for this to main_benchmark. Is this fine, or do we need to have the dependencies on master for the ci pipeline?
Also @Car-la-F uses seaborn for plotting. Since this makes life much easier than plain matplotlib, I would consider adding both. Please let us know how you feel about these packages.

FYI: the early wip code for snakemake is on benchmark_snakemake. Please be aware that this is incomplete and I might **force push** to this branch *anytime*. So you're welcome to have a look, but don't build on this. I'm going to open a PR towards the end of this week.

**Comment by smilesun** [comment]
sorry I kind of missed this thread among the simultaineous developments of many branches, yes, i think we can support more packages. those packages have been there for a long time, stable enough!
